### DISCUSSION TITLE
Tracking changed fields on users in more detail

### DISCUSSION THREAD

**Discussion by janhenckens** [craftcms/cms#13198] [Q&A] [answered]
We're working on a project where Craft will be partly used as a member management system for a medical society, and the client wants to track changes to user profiles.

I know there's the `changedfields` but that doesn't track old and new values I think?
Since users don't have revisions (and since those aren't permanent), would this we something we'd need to build out ourselves with a `element_before_save` event?

**Comment by brandonkelly** [comment, marked as answer]
Yeah currently users don’t have any sort of field value change tracking, so you would need to build it yourselves.

Here’s a starting point for how you could go about it:

```php
use craft\elements\User;
use craft\events\ModelEvent;
use yii\base\Event;

Event::on(User::class, User::EVENT_BEFORE_SAVE, function(ModelEvent $event) {
    if ($event->isNew) {
        return;
    }

    /** @var User $user */
    $user = $event->sender;
    $oldFieldValues = $user->getSerializedFieldValues();

    $user->on(User::EVENT_AFTER_SAVE, function() use ($user) {
        $newFieldValues = $user->getSerializedFieldValues();
        // compare them...
    });
});
```

**Reply by janhenckens** [reply]
Gotcha, thanks for the primer @brandonkelly 👍
